### DISCUSSION TITLE
Support psycopg `ClientCursor`

### DISCUSSION THREAD

**Discussion by CaselIT** [sqlalchemy/sqlalchemy#8978] [Ideas]
psycopg 3.1 added a  `ClientCursor` that pretty much behaves like psycopg2, doing client side parameter binding. docs at https://www.psycopg.org/psycopg3/docs/api/cursors.html#psycopg.ClientCursor

We may want to support it, since it would make migrating psycopg2 -> psycopg(3) easier.

This could probably be a dialect flag, since I think SQLAlchemy needs to remove the casting of all parameter that is added in psycopg3 and asyncpg that use server side parameter binding.

**Comment by zzzeek** [comment]
any server-side binding that doesnt work in SQLAlchemy (such as #8977) is a bug, that we will fix.

what other advantages are there to client side bindings?  why would someone select it (besides working around bugs that SQLAlchemy should fix).

my hesitancy is as always, to avoid supporting new variations that multiply our testing and support burden, especially if this is to be some dialect-level flag that is seldom used, but gets in the way when at some point we decide we want to use just one kind of binding.

See for example the debacle over mssql fast_executemany.

**Reply by CaselIT** [reply]
well if it was added to psycopg I guess they got some request for it.

In any case I agree and I've opened a discussion instead of an issue to judge the interest for it.
We can defer deciding on it until a later moment

**Reply by zzzeek** [reply]
does ClientCursor not use a prepared statement and therefore has better performance for SELECT ?

**Reply by CaselIT** [reply]
Don't know about perf, but it doesn't use prepared statementa (doesn't support them) from https://www.psycopg.org/psycopg3/docs/advanced/cursors.html#client-side-binding-cursors

**Reply by odelmarcelle** [reply]
I used a fairly complex statement to upsert a large number of ORM objects and return whether the rows were updated, inserted, or unchanged. The query was constructed around a call to `values(cols).data(data_tuples)`, where data_tuple could contain 10-100k tuples of rows to upsert. To get the number of inserted/upserted/no-oop rows, I wrapped the VALUES construct in a CTE.

The shape of the query is something like:
```
with input_rows as (
select ... from (values ( :param_1, :param_2),  ( :param_3, :param_4),  ( :param_5, :param_6), ....  )),
inserted_rows as (
insert into some_table (...) from input_rows on conflict do update ... where ...  returning ... )
select ... from inserted_rows union all select ... from input_rows
```

This worked fined in psycopg2 as the binding of the VALUES construct happened client-side. Using psycopg3, I'm hitting the limit of the number of parameters sent to the server (https://github.com/psycopg/psycopg/issues/620). It's not clear to me how I could modify that query to use sqlalchemy's executemany feature, since the call to `values()` expect a list of tuples as data. Using the `literal_binds` argument of `values()` is of no help as well, since I have some JSONB columns that won't compile.

For that specific case, it seems that the `ClientCursor` would be the solution to my problem. Otherwise, I'll have to revert to psycopg2 or chunk manually the parameters passed to `values(cols).data(data_tuples)`.

Happy to learn if my approach to the problem is not correct and that's why I'm hitting this issue to start with.

**Reply by CaselIT** [reply]
I think a PRs that adds options to the dialects to allow setting the cursor class for psycopg would be something that could be accepted.

**Reply by odelmarcelle** [reply]
Turns out it is already supported by passing `"cursor_factory": psycopg.ClientCursor` to `connect_args` while creating the engine. That solved my problem

**Reply by CaselIT** [reply]
Oh, did not know yoy could configure it like that! I'll mention it in the docs!

Thanks!

**Reply by CaselIT** [reply]
Documented it here: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5249
